crates/gpui/src/platform/linux/wayland/client.rs

```diff
@@ -163,6 +163,15 @@ impl WaylandClientStatePtr {
                 state.mouse_focused_window = Some(window);
             }
         }
+        if state.windows.is_empty() {
+            state
+                .common
+                .foreground_executor
+                .spawn(async {
+                    std::process::exit(0);
```

Comment:
We should use calloop to`LoopSignal::stop()` the main loop, so that the quit handler is fired before process exit.

Reply:
I tried doing that, but the process would seg fault.

Reply:
We should fix that seg fault then, it's important that the quit handler gets called so applications have a chance to clean things up.

Reply:
I do know we already have a seg fault on exit, so I don't think you need to fix it in this PR if it's pre-existing :)

Reply:
Weird thing is that on X11, `state.common.signal.stop()` is called, and it exits properly.

Reply:
Yeah, I think we're dropping things in the wrong order.

Reply:
I've debugged the crash and looks like the issue is with the `copypasta` crate, since `smithay-clipboard` (the underlying library which it uses on Wayland) isn't notified that we closed everything, so it crashes.
![image](https://github.com/zed-industries/zed/assets/71973804/b26d755a-ce9a-4d4e-afca-88405797288a)
![image](https://github.com/zed-industries/zed/assets/71973804/9c1989b3-3800-4234-8df4-68799f3e0da4)

Maybe we should switch to using `smithay-clipboard` directly since it basically is the same API? Any way, should I change this PR to stop the event loop and fix that issue in different one?

Reply:
Ahhhh, that makes a lot of sense. Yeah, let's fix the clipboard issue in a separate PR.

Reply:
Can we just drop the clipboard after `quit()` but before `run()` exits? Or do we need to drop them from within the running loop?

Reply:
I get this error when trying to drop it in the `state.windows.is_empty()` check
![image](https://github.com/zed-industries/zed/assets/71973804/412edf93-fc92-4f01-b9d3-1c59aa8ee082)

Reply:
I was actually thinking we could drop them in `WaylandClient::run()`

Reply:
Or the `platform.rs` run implementation could call a new method on `LinuxClient` after the `quit` callback is run, that gives the wayland platform a chance to drop the clipboards.

Reply:
Either way though, you'll need to put them in an option or something so you can take them out.